scripts/eslint/rules/mark-jsdoc-types-used.cjs

```diff
@@ -0,0 +1,102 @@
+    meta: {
+        docs: {
+            description: ``,
+        },
+        messages: {},
+        schema: [],
+        type: "problem",
+        fixable: "whitespace",
+    },
+    defaultOptions: [],
+
+    create(context) {
+        /** @type {(node: TSESTree.Node) => void} */
+        const checkProgram = (node) => {
+            const parserServices = ESLintUtils.getParserServices(context, /*allowWithoutFullTypeInformation*/ true);
+            const ast = parserServices.esTreeNodeToTSNodeMap.get(node);
+
+            // TODO(jakebailey): I almost guarantee this is overzealous and wrong in some way.
+
+            /**
+             * @param {ts.Node} node
+             */
+            function markIdentifiersUsed(node) {
+                if (ts.isIdentifier(node) && isBareIdentifier(node)) {
+                    context.markVariableAsUsed(node.text);
+                    return;
+                }
+                ts.forEachChild(node, markIdentifiersUsed);
+            }
+
+            /**
+             * @param {ts.Node} node
+             */
+            function visit(node) {
+                const jsDoc = ts.getJSDocTags(node);
+                for (const tag of jsDoc) {
+                    if (ts.isJSDocTypeTag(tag)) {
+                        markIdentifiersUsed(tag.typeExpression);
+                    }
+                    else if (ts.isJSDocTypedefTag(tag) || ts.isJSDocPropertyLikeTag(tag) || ts.isJSDocReturnTag(tag) || ts.isJSDocThrowsTag(tag)) {
+                        if (tag.typeExpression) {
+                            markIdentifiersUsed(tag.typeExpression);
+                        }
+                    }
+                    else if (ts.isJSDocTemplateTag(tag)) {
+                        if (tag.constraint) {
+                            markIdentifiersUsed(tag.constraint);
+                        }
+                        // tag.typeParameters?
+                    }
+                    else if (ts.isJSDocEnumTag(tag)) {
+                        markIdentifiersUsed(tag.typeExpression);
+                    }
+                    else if (ts.isJSDocUnknownTag(tag)) {
+                        // Ignore.
+                        // TODO(jakebailey): Something's wrong here, though, becuase we're getting @parameter tags as unknown. See new code in parser.
+                    }
+                    else if (ts.isJSDocOverrideTag(tag) || ts.isJSDocDeprecatedTag(tag)) {
+                        // Ignore.
+                    }
+                    else if (ts.isJSDocSeeTag(tag)) {
+                        markIdentifiersUsed(tag.tagName);
+                    }
+                    else {
+                        throw new Error(`Unexpected node kind: ${ts.SyntaxKind[tag.kind]} ${tag.getText()}}`);
+                    }
+                }
+                ts.forEachChild(node, visit);
+            }
+
+            visit(ast);
+        };
+
+        return {
+            Program: checkProgram,

```

Comment:
One problem I can see here is that you're accessing the comments from the program root.

Scopes are ofc scoped - so `context.markVariableAsUsed` is scoped too. So by doing all the work from the Program node you're accessing the top-level scope (eg the module scope) and thus will only ever mark top-level variables as used.

If you want to be able to handle nested scopes then we need to find the containing scope for the comment - which will require more effort.

Reply:
Yeah, this is what I was noting above; I do have another JSDoc rule which instead walks specifically only declarations and then asks for their JSDoc, but I wasn't sure if it was possible to have JSDoc hanging around somewhere else. e.g. I know in that other rule I'm definitely missing JSDoc on `EndOfFileToken` but for that rule it doesn't matter because it's about d.ts files and the bundler won't deal with it.

But, with this API, I'm not super sure how this scoping is supposed to work; a JSDoc node can go arbitrarily deep and introduce new scopes, which eslint is not going to know about because it's not seeing the full tree.

Reply:
Of course, this begs the question; how is the jsdoc plugin doing this? Or is it bugged too? Is this hopeless?

Reply:
It's a bit of a black box I think based on however [their parser package](https://github.com/es-joy/jsdoccomment) works and traverses jsdoc comments

https://github.com/gajus/eslint-plugin-jsdoc/blob/eed807edd9a14750ae4e16279cafeb27064ecd59/src/rules/noUndefinedTypes.js#L288

But essentially `if (type === 'JsdocTypeName') { mark_as_used(value) }`.
So it probably doesn't handle all these edge-cases you're thinking of?